doghouse/appengine/github.go

```diff
@@ -0,0 +1,294 @@
+	ctx := appengine.NewContext(r)
+
+	ok, token := g.token(r)
+	if !ok {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	ghcli := github.NewClient(NewAuthClient(ctx, urlfetch.Client(ctx).Transport, ts))
+
+	// /gh/{owner}/{repo}
+	paths := strings.Split(strings.Trim(r.URL.Path, "/"), "/")
+	switch len(paths) {
+	case 1:
+		g.handleTop(ctx, ghcli, w, r)
+	case 3:
+		g.handleRepo(ctx, ghcli, w, r, paths[1], paths[2])
+	default:
+		notfound(w)
+	}
+}
+
+func notfound(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusNotFound)
+	fmt.Fprintln(w, "404 Not Found")
+}
+
+func (g *GitHubHandler) handleTop(ctx context.Context, ghcli *github.Client, w http.ResponseWriter, r *http.Request) {
+	u, _, err := ghcli.Users.Get(ctx, "")
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintf(w, "cannot get GitHub authenticated user")
+		return
+	}
+
+	fmt.Fprintf(w, "Hi, %s!\n", u.GetName())
+
+	{
+		fmt.Fprintln(w, "")
+		ghcli, err := server.NewGitHubClient(ctx, &server.NewGitHubClientOption{
+			Client:        urlfetch.Client(ctx),
+			IntegrationID: g.integrationID,
+			PrivateKey:    g.privateKey,
+		})
+		if err != nil {
+			fmt.Fprintln(w, err)
+			return
+		}
+		app, _, err := ghcli.Apps.Get(ctx, "")
+		if err != nil {
+			fmt.Fprintln(w, err)
+		} else {
+			fmt.Fprintf(w, "%s app setting: %s\n", app.GetName(), app.GetHTMLURL())
+		}
+	}
+
+	{
+		fmt.Fprintln(w, "")
+		fmt.Fprintf(w, "Installation setting\n")
+		installations, _, err := ghcli.Apps.ListUserInstallations(ctx, nil)
+		if err != nil {
+			fmt.Fprintln(w, err)
+			return
+		}
+		for _, inst := range installations {
+			fmt.Fprintf(w, "\t%s: %s\n", inst.GetAccount().GetLogin(), inst.GetHTMLURL())
+		}
+	}
+}
+
+func (g *GitHubHandler) handleRepo(ctx context.Context, ghcli *github.Client, w http.ResponseWriter, r *http.Request, owner, repoName string) {
+	repo, _, err := ghcli.Repositories.Get(ctx, owner, repoName)
+	if err != nil {
+		if err, ok := err.(*github.ErrorResponse); ok {
+			if err.Response.StatusCode == http.StatusNotFound {
+				notfound(w)
+				return
+			}
+		}
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintf(w, "failed to get repo: %#v", err)
+		return
+	}
+
+	if !repo.GetPermissions()["push"] {
+		w.WriteHeader(http.StatusUnauthorized)
+		fmt.Fprintf(w, "You don't have write permission for %s.", repo.GetHTMLURL())
+		return
+	}
+
+	repoToken, err := server.GetOrGenerateRepoToken(ctx, g.repoTokenStore, repo.Owner.GetLogin(), repo.GetName(), repo.GetID())
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "failed to get repository token for %s.", repo.GetHTMLURL())
+		return
+	}
+
+	fmt.Fprintf(w, "%s repository token: %s", repo.GetFullName(), repoToken)
+}
+
+func NewAuthClient(ctx context.Context, base http.RoundTripper, token oauth2.TokenSource) *http.Client {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function NewAuthClient should have comment or be unexported